# Explore here

In [36]:
import pandas as pd
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.feature_selection import SelectKBest, f_regression
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.linear_model import Lasso
from pickle import dump

In [2]:
df = pd.read_csv("https://raw.githubusercontent.com/4GeeksAcademy/regularized-linear-regression-project-tutorial/main/demographic_health_data.csv", sep = ",")

save_path = r"D:\01A-TRABAJO\PYTHON\DATASCIENCE\EJERCICIOS\LRL\RLR\data\raw\raw_data.csv"

# We save the csv file
df.to_csv(save_path, index=False)

df.head()

fips  TOT_POP    0-9  0-9 y/o % of total pop  19-Oct  \
0  1001    55601   6787               12.206615    7637   
1  1003   218022  24757               11.355276   26913   
2  1005    24881   2732               10.980266    2960   
3  1007    22400   2456               10.964286    2596   
4  1009    57840   7095               12.266598    7570   

   10-19 y/o % of total pop  20-29  20-29 y/o % of total pop  30-39  \
0                 13.735364   6878                 12.370281   7089   
1                 12.344167  23579                 10.814964  25213   
2                 11.896628   3268                 13.134520   3201   
3                 11.589286   3029                 13.522321   3113   
4                 13.087828   6742                 11.656293   6884   

   30-39 y/o % of total pop  ...  COPD_number  diabetes_prevalence  \
0                 12.749771  ...         3644                 12.9   
1                 11.564429  ...        14692                 12.0   
2                 12.865239  ...         2373                 19.7   
3                 13.897321  ...         1789                 14.1   
4                 11.901798  ...         4661                 13.5   

   diabetes_Lower 95% CI  diabetes_Upper 95% CI  diabetes_number  \
0                   11.9                   13.8             5462   
1                   11.0                   13.1            20520   
2                   18.6                   20.6             3870   
3                   13.2                   14.9             2511   
4                   12.6                   14.5             6017   

   CKD_prevalence  CKD_Lower 95% CI  CKD_Upper 95% CI  CKD_number  \
0             3.1               2.9               3.3        1326   
1             3.2               3.0               3.5        5479   
2             4.5               4.2               4.8         887   
3             3.3               3.1               3.6         595   
4             3.4               3.2               3.7        1507   

   Urban_rural_code  
0                 3  
1                 4  
2                 6  
3                 2  
4                 2  

[5 rows x 108 columns]

we want to find out if there is any relationship between health resources and sociodemographic data

In [3]:
df.info

<bound method DataFrame.info of        fips  TOT_POP    0-9  0-9 y/o % of total pop  19-Oct  \
0      1001    55601   6787               12.206615    7637   
1      1003   218022  24757               11.355276   26913   
2      1005    24881   2732               10.980266    2960   
3      1007    22400   2456               10.964286    2596   
4      1009    57840   7095               12.266598    7570   
...     ...      ...    ...                     ...     ...   
3135  56037    43051   6104               14.178532    6326   
3136  56039    23081   2384               10.328842    2185   
3137  56041    20299   3121               15.375142    3205   
3138  56043     7885    858               10.881420    1113   
3139  56045     6967    780               11.195637     779   

      10-19 y/o % of total pop  20-29  20-29 y/o % of total pop  30-39  \
0                    13.735364   6878                 12.370281   7089   
1                    12.344167  23579                 10.814964

In [4]:
df.shape

(3140, 108)

In [5]:
print(df.columns.tolist()) 

['fips', 'TOT_POP', '0-9', '0-9 y/o % of total pop', '19-Oct', '10-19 y/o % of total pop', '20-29', '20-29 y/o % of total pop', '30-39', '30-39 y/o % of total pop', '40-49', '40-49 y/o % of total pop', '50-59', '50-59 y/o % of total pop', '60-69', '60-69 y/o % of total pop', '70-79', '70-79 y/o % of total pop', '80+', '80+ y/o % of total pop', 'White-alone pop', '% White-alone', 'Black-alone pop', '% Black-alone', 'Native American/American Indian-alone pop', '% NA/AI-alone', 'Asian-alone pop', '% Asian-alone', 'Hawaiian/Pacific Islander-alone pop', '% Hawaiian/PI-alone', 'Two or more races pop', '% Two or more races', 'POP_ESTIMATE_2018', 'N_POP_CHG_2018', 'GQ_ESTIMATES_2018', 'R_birth_2018', 'R_death_2018', 'R_NATURAL_INC_2018', 'R_INTERNATIONAL_MIG_2018', 'R_DOMESTIC_MIG_2018', 'R_NET_MIG_2018', 'Less than a high school diploma 2014-18', 'High school diploma only 2014-18', "Some college or associate's degree 2014-18", "Bachelor's degree or higher 2014-18", 'Percent of adults with les

Ver si hay duplicados, si hay valores nulos, si hay alguna columna con solo un valor, 

In [6]:
df.head()

fips  TOT_POP    0-9  0-9 y/o % of total pop  19-Oct  \
0  1001    55601   6787               12.206615    7637   
1  1003   218022  24757               11.355276   26913   
2  1005    24881   2732               10.980266    2960   
3  1007    22400   2456               10.964286    2596   
4  1009    57840   7095               12.266598    7570   

   10-19 y/o % of total pop  20-29  20-29 y/o % of total pop  30-39  \
0                 13.735364   6878                 12.370281   7089   
1                 12.344167  23579                 10.814964  25213   
2                 11.896628   3268                 13.134520   3201   
3                 11.589286   3029                 13.522321   3113   
4                 13.087828   6742                 11.656293   6884   

   30-39 y/o % of total pop  ...  COPD_number  diabetes_prevalence  \
0                 12.749771  ...         3644                 12.9   
1                 11.564429  ...        14692                 12.0   
2                 12.865239  ...         2373                 19.7   
3                 13.897321  ...         1789                 14.1   
4                 11.901798  ...         4661                 13.5   

   diabetes_Lower 95% CI  diabetes_Upper 95% CI  diabetes_number  \
0                   11.9                   13.8             5462   
1                   11.0                   13.1            20520   
2                   18.6                   20.6             3870   
3                   13.2                   14.9             2511   
4                   12.6                   14.5             6017   

   CKD_prevalence  CKD_Lower 95% CI  CKD_Upper 95% CI  CKD_number  \
0             3.1               2.9               3.3        1326   
1             3.2               3.0               3.5        5479   
2             4.5               4.2               4.8         887   
3             3.3               3.1               3.6         595   
4             3.4               3.2               3.7        1507   

   Urban_rural_code  
0                 3  
1                 4  
2                 6  
3                 2  
4                 2  

[5 rows x 108 columns]

In [7]:
print(df.nunique().to_string())

fips                                                                             3140
TOT_POP                                                                          3074
0-9                                                                              2723
0-9 y/o % of total pop                                                           3136
19-Oct                                                                           2743
10-19 y/o % of total pop                                                         3137
20-29                                                                            2698
20-29 y/o % of total pop                                                         3137
30-39                                                                            2708
30-39 y/o % of total pop                                                         3133
40-49                                                                            2693
40-49 y/o % of total pop                              

In [8]:
df.nunique()

fips                      3140
TOT_POP                   3074
0-9                       2723
0-9 y/o % of total pop    3136
19-Oct                    2743
                          ... 
CKD_prevalence              43
CKD_Lower 95% CI            39
CKD_Upper 95% CI            46
CKD_number                1894
Urban_rural_code             6
Length: 108, dtype: int64

In [9]:
df.drop_duplicates().reset_index(drop = True)

fips  TOT_POP    0-9  0-9 y/o % of total pop  19-Oct  \
0      1001    55601   6787               12.206615    7637   
1      1003   218022  24757               11.355276   26913   
2      1005    24881   2732               10.980266    2960   
3      1007    22400   2456               10.964286    2596   
4      1009    57840   7095               12.266598    7570   
...     ...      ...    ...                     ...     ...   
3135  56037    43051   6104               14.178532    6326   
3136  56039    23081   2384               10.328842    2185   
3137  56041    20299   3121               15.375142    3205   
3138  56043     7885    858               10.881420    1113   
3139  56045     6967    780               11.195637     779   

      10-19 y/o % of total pop  20-29  20-29 y/o % of total pop  30-39  \
0                    13.735364   6878                 12.370281   7089   
1                    12.344167  23579                 10.814964  25213   
2                    11.896628   3268                 13.134520   3201   
3                    11.589286   3029                 13.522321   3113   
4                    13.087828   6742                 11.656293   6884   
...                        ...    ...                       ...    ...   
3135                 14.694200   5359                 12.448027   6577   
3136                  9.466661   2967                 12.854729   4093   
3137                 15.788955   2153                 10.606434   2702   
3138                 14.115409    715                  9.067850    903   
3139                 11.181283    681                  9.774652    906   

      30-39 y/o % of total pop  ...  COPD_number  diabetes_prevalence  \
0                    12.749771  ...         3644                 12.9   
1                    11.564429  ...        14692                 12.0   
2                    12.865239  ...         2373                 19.7   
3                    13.897321  ...         1789                 14.1   
4                    11.901798  ...         4661                 13.5   
...                        ...  ...          ...                  ...   
3135                 15.277229  ...         2098                  8.9   
3136                 17.733200  ...          928                  7.2   
3137                 13.311001  ...         1163                 10.4   
3138                 11.452124  ...          506                 11.3   
3139                 13.004162  ...          480                 11.7   

      diabetes_Lower 95% CI  diabetes_Upper 95% CI  diabetes_number  \
0                      11.9                   13.8             5462   
1                      11.0                   13.1            20520   
2                      18.6                   20.6             3870   
3                      13.2                   14.9             2511   
4                      12.6                   14.5             6017   
...                     ...                    ...              ...   
3135                    8.3                    9.6             2834   
3136                    6.5                    8.0             1360   
3137                    9.5                   11.2             1500   
3138                   10.3                   12.1              686   
3139                   10.7                   12.7              644   

      CKD_prevalence  CKD_Lower 95% CI  CKD_Upper 95% CI  CKD_number  \
0                3.1               2.9               3.3        1326   
1                3.2               3.0               3.5        5479   
2                4.5               4.2               4.8         887   
3                3.3               3.1               3.6         595   
4                3.4               3.2               3.7        1507   
...              ...               ...               ...         ...   
3135             2.6               2.4               2.8         821   
3136             2.4               2.2               2.

To reach our goal which is to  find out if there is any relationship between health resources and sociodemographic data, we will begin by dropping the columns that provide geographical data.

In [10]:
df = df.drop(columns=['fips', 'COUNTY_NAME', 'STATE_NAME', 'STATE_FIPS', 'CNTY_FIPS'])

In [11]:
df.isnull().sum()

TOT_POP                     0
0-9                         0
0-9 y/o % of total pop      0
19-Oct                      0
10-19 y/o % of total pop    0
                           ..
CKD_prevalence              0
CKD_Lower 95% CI            0
CKD_Upper 95% CI            0
CKD_number                  0
Urban_rural_code            0
Length: 103, dtype: int64

In [12]:
df.shape

(3140, 103)

In [13]:
print(df.isnull().mean().mul(100).to_string())

TOT_POP                                                                          0.0
0-9                                                                              0.0
0-9 y/o % of total pop                                                           0.0
19-Oct                                                                           0.0
10-19 y/o % of total pop                                                         0.0
20-29                                                                            0.0
20-29 y/o % of total pop                                                         0.0
30-39                                                                            0.0
30-39 y/o % of total pop                                                         0.0
40-49                                                                            0.0
40-49 y/o % of total pop                                                         0.0
50-59                                                            

In [14]:
df.describe().T

count           mean            std   min  \
TOT_POP                   3140.0  104189.412420  333583.395432  88.0   
0-9                       3140.0   12740.302866   41807.301846   0.0   
0-9 y/o % of total pop    3140.0      11.871051       2.124081   0.0   
19-Oct                    3140.0   13367.976752   42284.392134   0.0   
10-19 y/o % of total pop  3140.0      12.694609       1.815044   0.0   
...                          ...            ...            ...   ...   
CKD_prevalence            3140.0       3.446242       0.568059   1.8   
CKD_Lower 95% CI          3140.0       3.207516       0.527740   1.7   
CKD_Upper 95% CI          3140.0       3.710478       0.613069   1.9   
CKD_number                3140.0    2466.234076    7730.422067   3.0   
Urban_rural_code          3140.0       4.635350       1.510447   1.0   

                                   25%           50%           75%  \
TOT_POP                   10963.250000  25800.500000  67913.000000   
0-9                        1280.500000   3057.000000   8097.000000   
0-9 y/o % of total pop       10.594639     11.802727     12.951840   
19-Oct                     1374.500000   3274.000000   8822.250000   
10-19 y/o % of total pop     11.674504     12.687422     13.659282   
...                                ...           ...           ...   
CKD_prevalence                3.100000      3.400000      3.800000   
CKD_Lower 95% CI              2.900000      3.200000      3.500000   
CKD_Upper 95% CI              3.300000      3.700000      4.100000   
CKD_number                  314.750000    718.000000   1776.250000   
Urban_rural_code              3.000000      5.000000      6.000000   

                                   max  
TOT_POP                   1.010552e+07  
0-9                       1.208253e+06  
0-9 y/o % of total pop    2.546068e+01  
19-Oct                    1.239139e+06  
10-19 y/o % of total pop  2.330437e+01  
...                                ...  
CKD_prevalence            6.200000e+00  
CKD_Lower 95% CI          5.800000e+00  
CKD_Upper 95% CI          6.600000e+00  
CKD_number                2.377660e+05  
Urban_rural_code          6.000000e+00  

[103 rows x 8 columns]

We will proceed to scale the data so it is all of it in the same level. 

In [15]:
data_types = df.dtypes
numeric_columns = [c for c in list(data_types[data_types != "object"].index) if c != "Heart disease_number"]

scaler = StandardScaler()
norm_features = scaler.fit_transform(df[numeric_columns])

# Create a new DataFrame with the scaled numerical variables
total_data_scal = pd.DataFrame(norm_features, index = df.index, columns = numeric_columns)
total_data_scal["Heart disease_number"] = df["Heart disease_number"]
total_data_scal.head()

TOT_POP       0-9  0-9 y/o % of total pop    19-Oct  \
0 -0.145679 -0.142421                0.158006 -0.135556   
1  0.341296  0.287476               -0.242861  0.320383   
2 -0.237785 -0.239429               -0.419441 -0.246181   
3 -0.245223 -0.246032               -0.426966 -0.254791   
4 -0.138966 -0.135053                0.186249 -0.137140   

   10-19 y/o % of total pop     20-29  20-29 y/o % of total pop     30-39  \
0                  0.573496 -0.153144                  0.027610 -0.139384   
1                 -0.193107  0.183774                 -0.469965  0.230620   
2                 -0.439718 -0.225971                  0.272104 -0.218759   
3                 -0.609076 -0.230792                  0.396168 -0.220555   
4                  0.216679 -0.155888                 -0.200808 -0.143570   

   30-39 y/o % of total pop     40-49  ...  diabetes_prevalence  \
0                  0.588469 -0.122010  ...            -0.063696   
1                 -0.110300  0.332513  ...            -0.394103   
2                  0.656538 -0.225725  ...             2.432709   
3                  1.264959 -0.226554  ...             0.376846   
4                  0.088582 -0.124495  ...             0.156575   

   diabetes_Lower 95% CI  diabetes_Upper 95% CI  diabetes_number  \
0              -0.071720              -0.089834        -0.129902   
1              -0.414900              -0.337677         0.376251   
2               2.483064               2.317776        -0.183415   
3               0.423984               0.299632        -0.229096   
4               0.195197               0.158008        -0.111247   

   CKD_prevalence  CKD_Lower 95% CI  CKD_Upper 95% CI  CKD_number  \
0       -0.609615         -0.582796         -0.669652   -0.147523   
1       -0.433549         -0.393279         -0.343373    0.389791   
2        1.855312          1.880929          1.777443   -0.204321   
3       -0.257483         -0.203761         -0.180233   -0.242100   
4       -0.081417         -0.014244         -0.017093   -0.124105   

   Urban_rural_code  Heart disease_number  
0         -1.082865                  3345  
1         -0.420704                 13414  
2          0.903618                  2159  
3         -1.745026                  1533  
4         -1.745026                  4101  

[5 rows x 103 columns]

In [16]:
X = total_data_scal.drop(columns=["Heart disease_number"])
y = total_data_scal["Heart disease_number"]
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 42)
train_indices = list(X_train.index)
test_indices = list(X_test.index)

In [17]:
# We select the 30% best columns for the model to reduce the number of columns
k = int(len(X_train.columns) * 0.3)
selection_model = SelectKBest(score_func = f_regression, k = k)
selection_model.fit(X_train, y_train)
ix = selection_model.get_support()

X_train_sel = pd.DataFrame(selection_model.transform(X_train), columns = X_train.columns.values[ix])
X_test_sel = pd.DataFrame(selection_model.transform(X_test), columns = X_test.columns.values[ix])

X_train_sel.head()

TOT_POP       0-9    19-Oct     20-29     30-39     40-49     50-59  \
0 -0.232556 -0.227731 -0.234284 -0.232951 -0.226353 -0.231316 -0.229599   
1 -0.158676 -0.178665 -0.180166 -0.188266 -0.175070 -0.161168 -0.134688   
2 -0.199114 -0.211128 -0.195138 -0.166782 -0.195036 -0.194045 -0.199725   
3 -0.036595 -0.037734 -0.017077 -0.057986 -0.052252 -0.033158 -0.020228   
4  0.090839  0.094680  0.101662  0.056721  0.042392  0.068095  0.101699   

      60-69     70-79       80+  ...  \
0 -0.233425 -0.234680 -0.234420  ...   
1 -0.105618 -0.119270 -0.091822  ...   
2 -0.219256 -0.222207 -0.205154  ...   
3 -0.032603 -0.023876 -0.046224  ...   
4  0.144664  0.140685  0.166099  ...   

   Family Medicine/General Practice Primary Care (2019)  \
0                                          -0.212643      
1                                          -0.116680      
2                                          -0.192263      
3                                           0.062458      
4                                           0.274818      

   Total Specialist Physicians (2019)  Total Population  Population Aged 60+  \
0                           -0.208590         -0.231195            -0.229737   
1                           -0.110850         -0.150293            -0.098866   
2                           -0.217668         -0.197005            -0.216056   
3                           -0.107888         -0.036940            -0.030034   
4                            0.194913          0.097767             0.161314   

   county_pop2018_18 and older  anycondition_number  Obesity_number  \
0                    -0.233171            -0.234370       -0.232975   
1                    -0.152859            -0.142645       -0.155304   
2                    -0.195125            -0.193205       -0.201976   
3                    -0.039882            -0.003321        0.006163   
4                     0.088485             0.165555        0.182740   

   COPD_number  diabetes_number  CKD_number  
0    -0.223516        -0.218609   -0.219329  
1    -0.110080        -0.131449   -0.130962  
2    -0.193106        -0.189197   -0.206391  
3    -0.007077        -0.047515   -0.045054  
4     0.265603         0.123040    0.132454  

[5 rows x 30 columns]

In [18]:
X_test_sel.head()

TOT_POP       0-9    19-Oct     20-29     30-39     40-49     50-59  \
0 -0.285286 -0.285362 -0.294836 -0.269566 -0.258568 -0.268541 -0.289649   
1  0.496553  0.433072  0.392170  0.544659  0.453677  0.391480  0.499744   
2 -0.260191 -0.255123 -0.265837 -0.246628 -0.234723 -0.240703 -0.264552   
3  0.039389  0.058341  0.059701 -0.018647  0.003236  0.030594  0.074401   
4  0.364272  0.281232  0.323623  0.525353  0.295990  0.288317  0.298029   

      60-69     70-79       80+  ...  \
0 -0.312989 -0.316763 -0.286734  ...   
1  0.668639  0.716353  0.476084  ...   
2 -0.289867 -0.289846 -0.290962  ...   
3  0.091003  0.060721  0.005012  ...   
4  0.461297  0.497760  0.351393  ...   

   Family Medicine/General Practice Primary Care (2019)  \
0                                          -0.303292      
1                                           0.853184      
2                                          -0.277451      
3                                           0.197267      
4                                           0.659217      

   Total Specialist Physicians (2019)  Total Population  Population Aged 60+  \
0                           -0.285225         -0.284324            -0.308211   
1                            0.424904          0.477184             0.620724   
2                           -0.261868         -0.257294            -0.287868   
3                            0.130719          0.036299             0.055281   
4                            0.305024          0.336581             0.423969   

   county_pop2018_18 and older  anycondition_number  Obesity_number  \
0                    -0.283698            -0.302439       -0.302292   
1                     0.517408             0.527360        0.516364   
2                    -0.259943            -0.249299       -0.259877   
3                     0.031494             0.035274        0.026108   
4                     0.390596             0.271127        0.273318   

   COPD_number  diabetes_number  CKD_number  
0    -0.324038        -0.276290   -0.281172  
1     0.443806         0.418504    0.454092  
2    -0.225107        -0.225970   -0.242229  
3     0.136643         0.003409    0.022352  
4     0.329669         0.256620    0.334804  

[5 rows x 30 columns]

In [19]:
X_test_sel.shape

(628, 30)

In [20]:
# We save the dataframes into csv files to have access to them easily.
X_train_sel["Heart disease_number"] = list(y_train)
X_test_sel["Heart disease_number"] = list(y_test)

X_train_sel.to_csv(r"D:\01A-TRABAJO\PYTHON\DATASCIENCE\EJERCICIOS\LRL\RLR\data\processed\clean_train.csv", index = False)
X_test_sel.to_csv(r"D:\01A-TRABAJO\PYTHON\DATASCIENCE\EJERCICIOS\LRL\RLR\data\processed\clean_test.csv", index = False)

In [21]:
final_data = pd.concat([X_train_sel, X_test_sel])
final_data.head()

TOT_POP       0-9    19-Oct     20-29     30-39     40-49     50-59  \
0 -0.232556 -0.227731 -0.234284 -0.232951 -0.226353 -0.231316 -0.229599   
1 -0.158676 -0.178665 -0.180166 -0.188266 -0.175070 -0.161168 -0.134688   
2 -0.199114 -0.211128 -0.195138 -0.166782 -0.195036 -0.194045 -0.199725   
3 -0.036595 -0.037734 -0.017077 -0.057986 -0.052252 -0.033158 -0.020228   
4  0.090839  0.094680  0.101662  0.056721  0.042392  0.068095  0.101699   

      60-69     70-79       80+  ...  Total Specialist Physicians (2019)  \
0 -0.233425 -0.234680 -0.234420  ...                           -0.208590   
1 -0.105618 -0.119270 -0.091822  ...                           -0.110850   
2 -0.219256 -0.222207 -0.205154  ...                           -0.217668   
3 -0.032603 -0.023876 -0.046224  ...                           -0.107888   
4  0.144664  0.140685  0.166099  ...                            0.194913   

   Total Population  Population Aged 60+  county_pop2018_18 and older  \
0         -0.231195            -0.229737                    -0.233171   
1         -0.150293            -0.098866                    -0.152859   
2         -0.197005            -0.216056                    -0.195125   
3         -0.036940            -0.030034                    -0.039882   
4          0.097767             0.161314                     0.088485   

   anycondition_number  Obesity_number  COPD_number  diabetes_number  \
0            -0.234370       -0.232975    -0.223516        -0.218609   
1            -0.142645       -0.155304    -0.110080        -0.131449   
2            -0.193205       -0.201976    -0.193106        -0.189197   
3            -0.003321        0.006163    -0.007077        -0.047515   
4             0.165555        0.182740     0.265603         0.123040   

   CKD_number  Heart disease_number  
0   -0.219329                  2072  
1   -0.130962                  3796  
2   -0.206391                  2222  
3   -0.045054                  5484  
4    0.132454                  8686  

[5 rows x 31 columns]

In [22]:
X_test_sel.head()

TOT_POP       0-9    19-Oct     20-29     30-39     40-49     50-59  \
0 -0.285286 -0.285362 -0.294836 -0.269566 -0.258568 -0.268541 -0.289649   
1  0.496553  0.433072  0.392170  0.544659  0.453677  0.391480  0.499744   
2 -0.260191 -0.255123 -0.265837 -0.246628 -0.234723 -0.240703 -0.264552   
3  0.039389  0.058341  0.059701 -0.018647  0.003236  0.030594  0.074401   
4  0.364272  0.281232  0.323623  0.525353  0.295990  0.288317  0.298029   

      60-69     70-79       80+  ...  Total Specialist Physicians (2019)  \
0 -0.312989 -0.316763 -0.286734  ...                           -0.285225   
1  0.668639  0.716353  0.476084  ...                            0.424904   
2 -0.289867 -0.289846 -0.290962  ...                           -0.261868   
3  0.091003  0.060721  0.005012  ...                            0.130719   
4  0.461297  0.497760  0.351393  ...                            0.305024   

   Total Population  Population Aged 60+  county_pop2018_18 and older  \
0         -0.284324            -0.308211                    -0.283698   
1          0.477184             0.620724                     0.517408   
2         -0.257294            -0.287868                    -0.259943   
3          0.036299             0.055281                     0.031494   
4          0.336581             0.423969                     0.390596   

   anycondition_number  Obesity_number  COPD_number  diabetes_number  \
0            -0.302439       -0.302292    -0.324038        -0.276290   
1             0.527360        0.516364     0.443806         0.418504   
2            -0.249299       -0.259877    -0.225107        -0.225970   
3             0.035274        0.026108     0.136643         0.003409   
4             0.271127        0.273318     0.329669         0.256620   

   CKD_number  Heart disease_number  
0   -0.281172                   698  
1    0.454092                 13982  
2   -0.242229                  1768  
3    0.022352                  6739  
4    0.334804                 11305  

[5 rows x 31 columns]

In [23]:
X_test_sel.tail()

TOT_POP       0-9    19-Oct     20-29     30-39     40-49     50-59  \
623 -0.226952 -0.219406 -0.225840 -0.230853 -0.212838 -0.215832 -0.222200   
624 -0.295174 -0.290625 -0.300371 -0.279935 -0.271226 -0.282276 -0.298756   
625 -0.189552 -0.205100 -0.196226 -0.191434 -0.186972 -0.180149 -0.183407   
626 -0.019742 -0.015725 -0.002270 -0.019696 -0.053089 -0.036540 -0.013111   
627 -0.175542 -0.181536 -0.163206 -0.200128 -0.179480 -0.157119 -0.169337   

        60-69     70-79       80+  ...  Total Specialist Physicians (2019)  \
623 -0.241248 -0.249174 -0.236658  ...                           -0.220435   
624 -0.319421 -0.327363 -0.310528  ...                           -0.290449   
625 -0.181676 -0.191001 -0.157152  ...                           -0.150490   
626  0.004195 -0.003466 -0.014968  ...                           -0.039921   
627 -0.172897 -0.164823 -0.163784  ...                           -0.208912   

     Total Population  Population Aged 60+  county_pop2018_18 and older  \
623         -0.226648            -0.243989                    -0.228757   
624         -0.293772            -0.320346                    -0.294902   
625         -0.183217            -0.177617                    -0.185384   
626         -0.013217            -0.000553                    -0.019660   
627         -0.188589            -0.181770                    -0.176718   

     anycondition_number  Obesity_number  COPD_number  diabetes_number  \
623            -0.247448       -0.242779    -0.292800        -0.242542   
624            -0.316619       -0.319322    -0.344715        -0.295920   
625            -0.178418       -0.175558    -0.174210        -0.178306   
626             0.065233        0.058235     0.088609         0.068519   
627            -0.180735       -0.178553    -0.231151        -0.181499   

     CKD_number  Heart disease_number  
623   -0.240418                  1432  
624   -0.300062                   375  
625   -0.176892                  2794  
626    0.026622                  6579  
627   -0.187113                  2486  

[5 rows x 31 columns]

We will start using the data we previously saved.

In [24]:
training_data = pd.read_csv(r"D:\01A-TRABAJO\PYTHON\DATASCIENCE\EJERCICIOS\LRL\RLR\data\processed\clean_train.csv")
testing_data = pd.read_csv(r"D:\01A-TRABAJO\PYTHON\DATASCIENCE\EJERCICIOS\LRL\RLR\data\processed\clean_test.csv")

training_data.head()

TOT_POP       0-9    19-Oct     20-29     30-39     40-49     50-59  \
0 -0.232556 -0.227731 -0.234284 -0.232951 -0.226353 -0.231316 -0.229599   
1 -0.158676 -0.178665 -0.180166 -0.188266 -0.175070 -0.161168 -0.134688   
2 -0.199114 -0.211128 -0.195138 -0.166782 -0.195036 -0.194045 -0.199725   
3 -0.036595 -0.037734 -0.017077 -0.057986 -0.052252 -0.033158 -0.020228   
4  0.090839  0.094680  0.101662  0.056721  0.042392  0.068095  0.101699   

      60-69     70-79       80+  ...  Total Specialist Physicians (2019)  \
0 -0.233425 -0.234680 -0.234420  ...                           -0.208590   
1 -0.105618 -0.119270 -0.091822  ...                           -0.110850   
2 -0.219256 -0.222207 -0.205154  ...                           -0.217668   
3 -0.032603 -0.023876 -0.046224  ...                           -0.107888   
4  0.144664  0.140685  0.166099  ...                            0.194913   

   Total Population  Population Aged 60+  county_pop2018_18 and older  \
0         -0.231195            -0.229737                    -0.233171   
1         -0.150293            -0.098866                    -0.152859   
2         -0.197005            -0.216056                    -0.195125   
3         -0.036940            -0.030034                    -0.039882   
4          0.097767             0.161314                     0.088485   

   anycondition_number  Obesity_number  COPD_number  diabetes_number  \
0            -0.234370       -0.232975    -0.223516        -0.218609   
1            -0.142645       -0.155304    -0.110080        -0.131449   
2            -0.193205       -0.201976    -0.193106        -0.189197   
3            -0.003321        0.006163    -0.007077        -0.047515   
4             0.165555        0.182740     0.265603         0.123040   

   CKD_number  Heart disease_number  
0   -0.219329                  2072  
1   -0.130962                  3796  
2   -0.206391                  2222  
3   -0.045054                  5484  
4    0.132454                  8686  

[5 rows x 31 columns]

In [25]:
X_train = training_data.drop(["Heart disease_number"], axis = 1)
y_train = training_data["Heart disease_number"]
X_test = testing_data.drop(["Heart disease_number"], axis = 1)
y_test = testing_data["Heart disease_number"]

In [26]:
model = LogisticRegression()
model.fit(X_train, y_train)

d:\01A-TRABAJO\PYTHON\DATASCIENCE\EJERCICIOS\LRL\.venv\Lib\site-packages\sklearn\linear_model\_logistic.py:469: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

In [31]:
print(f"Intercep (a): {model.intercept_}")
print(f"Coefficients: {model.coef_}")

Intercep (a): [-0.29993467 -0.30024311 -0.29872438 ... -2.16635915 -2.22339784
 -2.08955305]
Coefficients: [[-0.07148996 -0.06690703 -0.07314905 ... -0.11727391 -0.08235215
  -0.08309027]
 [-0.07162752 -0.06753208 -0.07376414 ... -0.11727572 -0.08222861
  -0.08277042]
 [-0.07133694 -0.06691124 -0.073165   ... -0.11701625 -0.08209558
  -0.08279051]
 ...
 [ 0.22158203  0.29340801  0.3539065  ...  0.32392114 -0.17237526
   0.32915966]
 [ 0.19231643  0.1405455   0.09283209 ...  0.22694262  0.12900066
   0.30218238]
 [ 0.2301378   0.09077815  0.12253551 ...  0.02113756  0.43275355
   0.28560255]]


In [32]:
y_pred = model.predict(X_test)
y_pred

array([ 1072,  8689,  1072,  8689,  8689,  1072,  1072,  1072,  1072,
        1072,  1072,  1072,  3376,  1072,  7128,  1072, 75432,  1072,
        1072,  1072,  1072,  1072,  1072,  1072, 40686,  1448,  1072,
        1072,  1072,  1072,  1072,  1072,  1072,  1072,  8506,  1072,
        1072,  1072,  1072,  1072,  1072,  8689,  1072,  1072,  1072,
        1072,  1072,  8506,  1072,  1072, 32863,  1072,  1072,  1072,
        8506,  1072, 53850,  1072,  1072,  1072,  7128,  1072,  1072,
        8506,  1072,  1448,  1072,  1072,  1072, 23641,  1072,  1072,
        1072,  1072,  1072,  1448, 43089,  1072,  1072,  1072,  1072,
        3376, 12367,  1072,  1072,  1072, 29483,  1072,  1072,  7128,
        1072,  1072,  1072,  1072,  1072,  1072,  1072,  1072,  8506,
        1072, 76128,  8689,  1072,  1072,  1072,  1072,  1072,  1072,
        8506, 25143,  1072,  8506, 25091,  3376,  1072, 16376,  1072,
        1072,  7128,  1072,  1072,  1072,  1072,  1072,  1072,  1072,
        3376,  1072,

In [33]:
print(f"MSE: {mean_squared_error(y_test, y_pred)}")
print(f"R2 Score: {r2_score(y_test, y_pred)}")

MSE: 24052608.58757962
R2 Score: 0.7251930590272113


In [39]:
alpha = 1.0
lasso_model = Lasso(alpha = alpha)

# Training the model
lasso_model.fit(X_train, y_train)

# We evaluate the performance of the model on the test data
score = lasso_model.score(X_test, y_test)
print("Coefficients:", lasso_model.coef_)
print("R2 score:", score)

Coefficients: [ 5155.0827115   1421.42082637 -2084.24244952  -917.37022925
  -748.56912195  4488.41072295   419.01872824 -1241.83330233
  3845.66255431  1023.42564785   950.10027466 -5342.50582078
   197.3976123   -916.83246987  -388.53091043  2778.01908669
   366.46885428    20.13956739 -1959.0387644    309.89166549
  -349.85763302  -548.28239293 -3107.34555874  3278.41020291
   288.32110373  1702.45185306  3013.59232337  5402.53356674
   771.1844608  -2078.66459212]
R2 score: 0.998159015238823


d:\01A-TRABAJO\PYTHON\DATASCIENCE\EJERCICIOS\LRL\.venv\Lib\site-packages\sklearn\linear_model\_coordinate_descent.py:697: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 4.569e+08, tolerance: 7.097e+07
  model = cd_fast.enet_coordinate_descent(


Before Optimisation: R2 Score: 0.7251930590272113

AFTER OPTIMISATION: R2 score: 0.998159015238823

We improved

In [37]:
dump(lasso_model, open(r"D:\01A-TRABAJO\PYTHON\DATASCIENCE\EJERCICIOS\LRL\RLR\models\model.sav", "wb"))